cupyx/scipy/sparse/linalg/solve.py

```diff
@@ -65,3 +66,237 @@ def lsqr(A, b):
+        A = cupyx.scipy.sparse.csr_matrix(A)
+    util._assert_nd_squareness(A)
+    util._assert_cupy_array(b)
+    n = A.shape[0]
+    if b.ndim != 1 or len(b) != n:
+        raise ValueError('b must be 1-d array whose size is same as A')
+
+    # support float32, float64, complex64, and complex128
+    if A.dtype.char in 'fdFD':
+        dtype = A.dtype.char
+    else:
+        dtype = numpy.promote_types(A.dtype.char, 'f').char
+
+    # Calculate initial residual
+    x0 = x0 if x0 is not None else cupy.zeros(n, dtype=dtype)
+    r = cupy.cusparse.spmv(A, x0, b, -1, 1)
+    p, q = cupy.zeros_like(r), cupy.zeros_like(r)
+    r_tilde = r.copy()
+
+    # Main BiCGSTAB loop
+
+    i = 0
+    rho = alpha = omega = info = 1
+
+    x = x0.copy()
+    tol = cupy.linalg.norm(b) * tol
+    atol = 0 if atol == 'legacy' else atol
+
+    # preconditioner
+    M, M_info = (_identity, None) if M is None else M
+
+    while i < maxiter:
+        i += 1
+        prev_rho = rho
+        rho = cupy.dot(r_tilde.conj(), r)
+        if rho == 0:
+            info = -10
+            break
+        beta = rho / prev_rho * alpha / omega
+        p = r + beta * (p - omega * q)
+
+        # Alpha step
+        phat = M(p)  # apply preconditioner
+
+        q = cupy.cusparse.spmv(A, phat)
+        alpha = rho / cupy.dot(r_tilde.conj(), q)
+        s = r - alpha * q
+
+        residual = cupy.linalg.norm(s)
+        if residual < tol or residual < atol:
+            x += alpha * phat
+            info = 0
+            break
+
+        # Omega step
+        shat = M(s)  # apply preconditioner
+
+        t = cupy.cusparse.spmv(A, shat)
+        omega = cupy.dot(cupy.conj(t), s) / cupy.square(cupy.linalg.norm(t))
+        if omega == 0:
+            info = -11
+            break
+
+        x += alpha * phat + omega * shat
+        r = s - omega * t
+
+        residual = cupy.linalg.norm(r)
+        if callback is not None:
+            callback(x, i, r)
+        if residual < tol or residual < atol:
+            info = 0
+            break
+
+    _destroy_ilu(M_info)
+
+    return x, info
+
+
+def spilu(A, drop_tol=1e-4):
+    """Compute an incomplete LU decomposition for a sparse, square matrix.
+
+    The resulting object is an approximation to the inverse of A.
+    See https://docs.nvidia.com/cuda/cusparse/#csrilu02_solve.
+
+    Args:
+        A (cupy.ndarray or cupyx.scipy.sparse.csr_matrix): The input matrix
+            with dimension ``(N, N)``
+        drop_tol (float, optional): Drop tolerance (0 <= tol <= 1) for an
+            incomplete LU decomposition. (default: 1e-4)
+            [Currently a dummy param]
+
+    Returns:
+        callable:
+            A function x = M(b) that approximates a linear solution for Ax = b
+        tuple:
+            the info objects that need to be
+
+    .. seealso:: :func:`scipy.sparse.linalg.spilu`
+    """
+    # support float32, float64, complex64, and complex128
+    handle = device.get_cusolver_sp_handle()
```

Comment:
You need to use `device.get_cusparse_handle()` as cuSparse is used in this routine..